Regions/neg adj3

oelbert opened the pull request:
## Purpose

Refactoring neg_adj3 to improve performance by removing Python code between stencil calls and by adding 2D fields 

## Code changes:

- `fillq` now uses 2d fields to accumulate sums
- Renamed internal stencil variables for clarity and consistency with the rest of the code
- Pulled Python code into stencils to allow full fusion
- Added a docstring to neg_adj3.compute()

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -1,177 +1,151 @@
-    qs: sd,
-    qi: sd,
-    qg: sd,
+    qvapor: sd,
```
All occurrences of `sd` can now be replaced with `FloatField`.

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -1,177 +1,151 @@
             q = q + dq / dp
     with computation(BACKWARD), interval(...):
-        if q > 0. and sum1 >= 1e-12 and sum2 > 0:
+        if q > 0.0 and sum1 >= 1e-12 and sum2 > 0:
```
Nice touch, `0.` always looks incomplete to me...

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -1,177 +1,151 @@
     with computation(FORWARD), interval(...):
         if q > 0:
-            sum1 = sum1[0, 0, -1] + q * dp
+            sum1 = sum1 + q * dp
```
Love that this works now...

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -1,177 +1,151 @@
-# need sum1, sum2 to be an accumulating floats
 @gtstencil()
-def fillq(q:sd, dp:sd):
+def fillq(q: FloatField, dp: FloatField, sum1: FloatFieldIJ, sum2: FloatFieldIJ):
```
So awesome you were able to stencil-ize this.

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -1,177 +1,151 @@
+from fv3core.utils.typing import FloatField, FloatFieldIJ
 
 
 sd = utils.sd
```
`sd` can be removed in favor of `FloatField`.

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -1,177 +1,151 @@
 @gtscript.function
-def fix_negative_ice(qv, qi, qs, qg, qr, ql, pt, lcpk, icpk, dq):
-    qsum = qi + qs
+def fix_negative_ice(qvapor, qice, qsnow, qgraupel, qrain, qliquid, pt, lcpk, icpk, dq):
```
Type annotations are nice on `gtscript.function`'s too, but not required.

oelbert replied:
Would you object to me not including them this PR? I think we'll be revisiting this file a bit in the future

eddie-c-davis replied:
No objection at all!

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -206,99 +180,131 @@ def fix_neg_cloud(dp: sd, qcld: sd):
 
 # Nonstencil code for reference:
-def fix_water_vapor_nonstencil(grid, qv, dp):
+def fix_water_vapor_nonstencil(grid, qvapor, dp):
```
Should we comment the non-stencil code?

oelbert replied:
Oops! I'd also be ok with just removing it at this point, though

eddie-c-davis replied:
Sounds good, we can save that for the next round too :)

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -206,99 +180,131 @@ def fix_neg_cloud(dp: sd, qcld: sd):
 
 
 def compute(qvapor, qliquid, qrain, qsnow, qice, qgraupel, qcld, pt, delp, delz, peln):
+    """
```
Docstrings, well done!

eddie-c-davis commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -322,28 +328,32 @@ def compute(qvapor, qliquid, qrain, qsnow, qice, qgraupel, qcld, pt, delp, delz,
         origin=grid.compute_origin(),
         domain=grid.domain_shape_compute(),
     )
-    qvapor[:, :, grid.npz] = upper_fix[:, :, 0]
```
Way to remove that numpy code!

eddie-c-davis reviewed (changes requested):
Excellent work! Just a couple of suggestions related to the typing.

eddie-c-davis reviewed (approved):
Looks great!

jdahm commented on fv3core/stencils/neg_adj3.py:
```diff
@@ -1,184 +1,157 @@
-        elif qs < 0.0:
-            qs = 0.0
-            qi = qsum
+        if qice < 0.0:
+            qice = 0.0
+            qsnow = qsum
+        elif qsnow < 0.0:
+            qsnow = 0.0
+            qice = qsum
```
Great work on the clearer variable names!

oelbert commented:
Purpose
Refactoring neg_adj3 to improve performance by removing Python code between stencil calls and by adding 2D fields

Code changes:
- fillq now uses 2d fields to accumulate sums
- Renamed internal stencil variables for clarity and consistency with the rest of the code
- Pulled Python code into stencils to allow full fusion
- Added a docstring to neg_adj3.compute()
